From 66307c48f577bff6a2db9834a1fded12b012ac50 Mon Sep 17 00:00:00 2001
From: Xiaolei Wang <xiaolei.wang@windriver.com>
Date: Wed, 24 Jun 2020 14:08:55 +0800
Subject: [PATCH] usb: chipidea: host: Illegal memory accessed, hcd has been
 released

Illegal memory accessed, hcd has been released. This is a problem
caused by commit 07e8dbe5584b8146c36ea9fb828f1f2ee79525e8 from
https://source.codeaurora.org/external/imx/linux-imx.git

BUG: KASAN: use-after-free in host_stop+0xd0/0x188
Read of size 1 at addr ffff80006351b331 by task kworker/u8:3/251

CPU: 2 PID: 251 Comm: kworker/u8:3 Tainted: G      D
5.2.43-yocto-standard #1
Hardware name: FSL i.MX8MM EVK board (DT)
Workqueue: ci_otg ci_otg_work
Call trace:
 dump_backtrace+0x0/0x1e0
 show_stack+0x24/0x30
 dump_stack+0xbc/0xf4
 print_address_description+0x6c/0x264
 __kasan_report+0x150/0x198
 kasan_report+0xc/0x18
 __asan_load1+0x4c/0x58
 host_stop+0xd0/0x188
 ci_handle_id_switch+0x9c/0x308
 ci_otg_work+0xc4/0x198
 process_one_work+0x3d8/0x8a8
 worker_thread+0x80/0x648
 kthread+0x180/0x1c0
 ret_from_fork+0x10/0x1c

Allocated by task 254:
 __kasan_kmalloc.isra.0.part.0+0x40/0xd8
 __kasan_kmalloc.isra.0+0xb4/0xd0
 kasan_kmalloc+0xc/0x18
 __kmalloc+0xb0/0x328
 __usb_create_hcd+0x48/0x390
 host_start+0x70/0x398
 ci_handle_id_switch+0x138/0x308
 ci_otg_work+0xc4/0x198
 process_one_work+0x3d8/0x8a8
 worker_thread+0x80/0x648
 kthread+0x180/0x1c0
 ret_from_fork+0x10/0x1c

Freed by task 251:
 __kasan_slab_free+0x110/0x210
 kasan_slab_free+0x10/0x18
 kfree+0x8c/0x2d0
 usb_put_hcd+0x84/0xb8
 host_stop+0x98/0x188
 ci_handle_id_switch+0x9c/0x308
 ci_otg_work+0xc4/0x198
 process_one_work+0x3d8/0x8a8
 worker_thread+0x80/0x648
 kthread+0x180/0x1c0
 ret_from_fork+0x10/0x1c

The buggy address belongs to the object at ffff80006351b300
 which belongs to the cache kmalloc-2k of size 2048
The buggy address is located 49 bytes inside of
 2048-byte region [ffff80006351b300, ffff80006351bb00)
The buggy address belongs to the page:
page:ffff7e00018d4600 refcount:1 mapcount:0 mapping:ffff800068003400
index:0xffff80006351f700 compound_mapcount: 0
flags: 0x10200(slab|head)
raw: 0000000000010200 ffff7e0001856c00 0000000200000002 ffff800068003400
raw: ffff80006351f700 00000000800f000c 00000001ffffffff 0000000000000000
page dumped because: kasan: bad access detected

Memory state around the buggy address:
 ffff80006351b200: fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb
 ffff80006351b280: fc fc fc fc fc fc fc fc fc fc fc fc fc fc fc fc
>ffff80006351b300: fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb
                                     ^
 ffff80006351b380: fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb
 ffff80006351b400: fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb
==================================================================

Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 drivers/usb/chipidea/host.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/drivers/usb/chipidea/host.c b/drivers/usb/chipidea/host.c
index c9961d592277..f48fbe979be3 100644
--- a/drivers/usb/chipidea/host.c
+++ b/drivers/usb/chipidea/host.c
@@ -364,8 +364,6 @@ static void host_stop(struct ci_hdrc *ci)
 		if (ci->platdata->reg_vbus && !ci_otg_is_fsm_mode(ci) &&
 			(ci->platdata->flags & CI_HDRC_TURN_VBUS_EARLY_ON))
 				regulator_disable(ci->platdata->reg_vbus);
-		if (hcd->self.is_b_host)
-			hcd->self.is_b_host = 0;
 	}
 	ci->hcd = NULL;
 	ci->otg.host = NULL;
-- 
2.17.1

